src/persist-client/src/read.rs

```diff
@@ -579,89 +707,51 @@ where
-                            // contain updates we've already emitted. For example, we
-                            // emitted `[1, 2)` and then compaction combined that batch
-                            // with a `[2, 3)` batch into a new `[1, 3)` batch. If this
-                            // happens, we just need to filter out anything < the
-                            // frontier. This frontier was the upper of the last batch
-                            // (and thus exclusive) so for the == case, we still emit.
-                            if !until.less_equal(&t) {
-                                return;
-                            }
-                        }
-                        HollowBatchReaderMetadata::Snapshot { as_of } => {
-                            // This time is covered by a listen
-                            if as_of.less_than(&t) {
-                                return;
-                            }
-                            t.advance_by(as_of.borrow())
-                        }
-                    }
-
-                    let k = self.metrics.codecs.key.decode(|| K::decode(k));
-                    let v = self.metrics.codecs.val.decode(|| V::decode(v));
-                    let d = D::decode(d);
-                    updates.push(((k, v), t, d));
-                },
-            )
-            .await;
+    /// Tracks that the `ReadHandle`'s machine's current `SeqNo` is being
+    /// "leased out" to a `ReaderEncirchedHollowBatch`, and cannot be garbage
+    /// collected until its lease has been dropped (`Self::drop_seqno_lease`).
+    pub fn lease_seqno(&mut self) -> SeqNo {
+        let lease = self.machine.seqno();
+        match self.leased_seqnos.last_mut() {
+            Some((seqno, leases)) if seqno == &lease => *leases += 1,
+            _ => {
+                // Sorted invariant maintained
+                debug_assert!(
+                    self.leased_seqnos.iter().all(|(seqno, _)| seqno < &lease),
+                    "leased_seqnos must remain sorted"
+                );
+                self.leased_seqnos.push((lease, 1));
+            }
         }
+        lease
+    }
 
-        // TODO: This is potentially suprising for the `ReadHandle` to manage.
-        // Instead, we likely want to add a struct akin to a `Listen`
-        // (BatchStreamFetcher?) that wraps the `ReadHandle` and `fetch_batch`
-        // method is only available on that (think the pattern that Listen has
-        // now).
-        if let HollowBatchReaderMetadata::Listen { since, .. } = batch.reader_metadata {
-            self.maybe_downgrade_since(&since).await;
+    /// Tracks that a `SeqNo` lease has been returned and can be dropped. Once
+    /// a `SeqNo` has no more outstanding leases, it can be removed, and
+    /// `Self::downgrade_since` no longer needs to prevent it from being
+    /// garbage collected.
+    pub fn drop_seqno_lease(&mut self, leased_seqno: SeqNo) {
```

Comment:
this protocol seems easy to get wrong. a couple things I think might help:
- add a newtype wrapper around `SeqNo`, perhaps `LeasedSeqNo` that is accepted by this method
- instead of the `let leased_seqno = batch.seqno(); fetcher.fetch_batch(batch); read.drop_seqno_lease(leased_seqno);` pattern, can we also return the `LeasedSeqNo` from `fetch_batch`? that way, the compiler helps makes you aware of the whole LeasedSeqNo system

Reply:
Thinking about this some more: I think we should not even expose the fact that there are these `SeqNos` and leases to the user of the API. All that the user cares about is a) I get batches, and I'm guaranteed to be able to read them b) I need to acknowledge to the issuer of the batch when I have read it, so that it can do any internal cleanup/compaction logic.

Seems we could introduce a newtype `ConsumedBatch(SeqNo)` or `AcknoledgedBatch(SeqNo)`. `BatchFetcher::fetch_batch(batch)` could return a tuple `(<updates or whatnot>, AcknowledgedBatch)`. And we then have a method `mark_batch_acknoledged()` or `mark_batch_consumed()` on the `ReadHandle`.

That way, we don't leak the internal details in the API. That might seem pedantic but it a) makes us more future proof, b) clearly documents/messages the intent of the API.

I just like encapsulation. :sweat_smile:

Reply:
Done in commit w/ message `persist: add encapsulation around SeqNo leases`